qiskit/transpiler/coupling.py

```diff
@@ -463,20 +463,38 @@ def connected_components(self) -> List["CouplingMap"]:
 
         will print ``3`` as index ``0`` in the second component is qubit 3 in the original cmap.
 
+        Args:
+            target (Target): If used the provided :class:`~.Target` will be used to filter any
+                qubits that do not have any supported operations so that those qubits
+                will not be included as a connected component in the output list. When not
+                specified these qubits without any supported operations will be included in
+                the output as a connected component of a single qubit.
+
         Returns:
             list: A list of :class:`~.CouplingMap` objects for each connected
                 components. The order of this list is deterministic but
                 implementation specific and shouldn't be relied upon as
                 part of the API.
         """
+        graph = self.graph
+        if target is not None:
+            qargs = target.qargs
+            graph = self.graph.copy()
+            for index in graph.node_indices():
+                for qargs in target.qargs:
+                    if index in qargs:
+                        break
+                else:
+                    graph.remove_node(index)
```

Comment:
Or this. It might be more efficient. It's compact. It avoids explicit iteration and removes some dups with sets. But it allocates a set for `graph.node_indices`.
```suggestion
            graph = self.graph.copy()
            qarg_nodes = {v for vs in target.qargs for v in vs}
            graph_nodes = set(graph.node_indices())
            graph.remove_nodes_from(list(graph_nodes.difference(qarg_nodes)))
```